### ISSUE TITLE
prop assertions

### ISSUE THREAD

**Issue by yyx990803** [yyx990803/vue#920]
It'd be nice for components to be able to check the props it receives, and this makes reusable components more robust. e.g.

``` js
Vue.component('example', {
  props: [
    { name: 'on-something', type: Function },
    { name: 'some-prop', type: Boolean, required: true }
  ]
})
```

This could be backwards-compatible and totally opt-in.

**Comment by kazupon** [comment]
Wow ReactJS-like !!
Looks pretty good.
I'm prefer that.

**Comment by azamat-sharapov** [comment]
Interesting. Could you give some real world use from this? Looks cool to me, but can't think of some way using it :)

**Comment by OEvgeny** [comment]
Looks good. What if required prop is not provided?

**Comment by yyx990803** [comment]
This is mostly intended for shipping standalone, reusable components. These props are essentially the API exposed by the component. If a required prop is not provided or a type mismatches, Vue will throw a warning for you.

**Comment by OEvgeny** [comment]
Thank you, Evan. I think it is a good addition to current component system.

**Comment by yyx990803** [comment]
Shipped in 0.12.1
